src/terms.rs

```diff
@@ -0,0 +1,529 @@
+        }
+    }
+
+    fn canonicalize_rules(&mut self, fdecl: &NormFunctionDecl) -> Vec<Command> {
+        let types = self.type_info().lookup_user_func(fdecl.name).unwrap();
+        let has_eq_type = types.output.is_eq_sort() || types.input.iter().any(|s| s.is_eq_sort());
+
+        if !has_eq_type {
+            return vec![];
+        }
+
+        let op_name = fdecl.name;
+
+        let view_name = if types.is_datatype {
+            self.view_name(op_name)
+        } else {
+            fdecl.name
+        };
+        let child = |i| format!("c{i}_");
+        let child_parent = |myself: &Self, i| {
+            #[allow(clippy::iter_nth)]
+            let child_t: ArcSort = types.input.iter().nth(i).unwrap().clone();
+            myself
+                .wrap_parent_or_rebuild(child(i), child_t)
+                .unwrap_or_else(|| child(i))
+        };
+        let children = format!(
+            "{}",
+            ListDisplay(
+                (0..fdecl.schema.input.len()).map(child).collect::<Vec<_>>(),
+                " "
+            )
+        );
+        let children_updated = format!(
+            "{}",
+            ListDisplay(
+                (0..fdecl.schema.input.len())
+                    .map(|v| child_parent(self, v))
+                    .collect::<Vec<_>>(),
+                " "
+            )
+        );
+        let lhs_updated = self
+            .wrap_parent_or_rebuild("lhs".to_string(), types.output.clone())
+            .unwrap_or_else(|| "lhs".to_string());
+
+        let mut res = vec![];
+        // This rule updates each row of the table,
+        // doing canonicalization
+        // (and through the view merge function, rebuilding)
+
+        // Actually, we do one rule per column for
+        // efficiency- joining on one column at a time
+        // is more efficient
+        for i in 0..(fdecl.schema.input.len() + 1) {
+            let var_type = if i == fdecl.schema.input.len() {
+                types.output.clone()
+            } else {
+                types.input[i].clone()
+            };
+            let current = if i == fdecl.schema.input.len() {
+                "lhs".to_string()
+            } else {
+                child(i)
+            };
+            let current_updated = if i == fdecl.schema.input.len() {
+                lhs_updated.clone()
+            } else {
+                child_parent(self, i)
+            };
+            let new_lhs = self.fresh_var();
+            let new_term_for_view = if types.output.is_eq_sort() {
+                format!(
+                    "(let {new_lhs} ({op_name} {children_updated}))
+                         {}",
+                    self.union(
+                        types.output.name(),
+                        &new_lhs.to_string(),
+                        lhs_updated.as_str(),
+                    )
+                )
+            } else {
+                "".to_string()
+            };
+            if var_type.is_eq_sort() {
```

Comment:
Should it be `var_type.is_eq_sort() || var_type.is_eq_container_sort()`?

Reply:
Oof good catch